src/librustc_privacy/lib.rs

```diff
@@ -897,10 +897,11 @@ impl<'a, 'tcx, 'v> Visitor<'v> for PrivacyVisitor<'a, 'tcx> {
             }
             ast::ExprStruct(_, ref fields, _) => {
                 match ty::expr_ty(self.tcx, expr).sty {
-                    ty::ty_struct(id, _) => {
-                        for field in &(*fields) {
-                            self.check_field(expr.span, id,
-                                             NamedField(field.ident.node));
+                    ty::ty_struct(ctor_id, _) => {
+                        let all_fields = ty::lookup_struct_fields(self.tcx, ctor_id);
```

Comment:
Might be worth a comment (maybe with a pointer to the RFC), clarifying that 

1. The fields that are explicitly mentioned must be public
2. The rest must also be public as described in RFC #xyz and issue #xyz